File: ex/lib/gralkor/client/in_memory.ex
defmodule Gralkor.Client.InMemory do
  @moduledoc """
  In-memory twin of `Gralkor.Client` for tests.

  Configure responses per operation before exercising the code under test;
  inspect the calls that were made after. Operations with no configured
  response return `{:error, :not_configured}` — tests should set every
  response they expect the code under test to hit.
  """

  @behaviour Gralkor.Client

  use GenServer

  # ── Test API ────────────────────────────────────────────────

  def start_link(_opts \\ []),
    do: GenServer.start_link(__MODULE__, :ok, name: __MODULE__)

  @doc "Reset all state — call in `setup`."
  def reset, do: GenServer.call(__MODULE__, :reset)

  @doc "Set the response for the next (and all subsequent) `recall/3` calls."
  def set_recall(response), do: GenServer.call(__MODULE__, {:set, :recall, response})

  @doc "Set the response for the next (and all subsequent) `capture/3` calls."
  def set_capture(response), do: GenServer.call(__MODULE__, {:set, :capture, response})

  @doc "Set the response for the next (and all subsequent) `memory_search/3` calls."
  def set_memory_search(response), do: GenServer.call(__MODULE__, {:set, :memory_search, response})

  @doc "Set the response for the next (and all subsequent) `memory_add/3` calls."
  def set_memory_add(response), do: GenServer.call(__MODULE__, {:set, :memory_add, response})

  @doc "Set the response for the next (and all subsequent) `end_session/1` calls."
  def set_end_session(response), do: GenServer.call(__MODULE__, {:set, :end_session, response})

  @doc "Set the response for the next (and all subsequent) `health_check/0` calls."
  def set_health(response), do: GenServer.call(__MODULE__, {:set, :health_check, response})

  def recalls, do: GenServer.call(__MODULE__, {:calls, :recall})
  def captures, do: GenServer.call(__MODULE__, {:calls, :capture})
  def searches, do: GenServer.call(__MODULE__, {:calls, :memory_search})
  def adds, do: GenServer.call(__MODULE__, {:calls, :memory_add})
  def end_sessions, do: GenServer.call(__MODULE__, {:calls, :end_session})
  def health_checks, do: GenServer.call(__MODULE__, {:calls, :health_check})

  # ── Client behaviour ────────────────────────────────────────

  @impl Gralkor.Client
  def recall(group_id, session_id, query),
    do: GenServer.call(__MODULE__, {:call, :recall, [group_id, session_id, query]})

  @impl Gralkor.Client
  def capture(session_id, group_id, turn),
    do: GenServer.call(__MODULE__, {:call, :capture, [session_id, group_id, turn]})

  @impl Gralkor.Client
  def memory_search(group_id, session_id, query),
    do: GenServer.call(__MODULE__, {:call, :memory_search, [group_id, session_id, query]})

  @impl Gralkor.Client
  def memory_add(group_id, content, source),
    do: GenServer.call(__MODULE__, {:call, :memory_add, [group_id, content, source]})

  @impl Gralkor.Client
  def end_session(session_id),
    do: GenServer.call(__MODULE__, {:call, :end_session, [session_id]})

  @impl Gralkor.Client
  def health_check,
    do: GenServer.call(__MODULE__, {:call, :health_check, []})

  # ── GenServer ──────────────────────────────────────────────

  @impl GenServer
  def init(:ok), do: {:ok, empty_state()}

  @impl GenServer
  def handle_call(:reset, _from, _state), do: {:reply, :ok, empty_state()}

  def handle_call({:set, op, response}, _from, state),
    do: {:reply, :ok, put_in(state.responses[op], response)}

  def handle_call({:calls, op}, _from, state),
    do: {:reply, Enum.reverse(Map.get(state.calls, op, [])), state}

  def handle_call({:call, op, args}, _from, state) do
    state = update_in(state.calls[op], &[args | &1 || []])
    response = Map.get(state.responses, op, {:error, :not_configured})
    {:reply, response, state}
  end

  defp empty_state, do: %{responses: %{}, calls: %{}}
end

File: ts/src/client/in-memory.ts
import type { GralkorClient, Message, Result } from "../client.js";

type Op =
  | "recall"
  | "capture"
  | "endSession"
  | "memorySearch"
  | "memoryAdd"
  | "healthCheck"
  | "buildIndices"
  | "buildCommunities";

/**
 * In-memory twin of {@link GralkorClient} for tests.
 *
 * Configure canned responses with {@link setResponse} before exercising the
 * code under test; inspect calls via the `recalls`, `captures`, `searches`,
 * `adds`, `endSessions`, `healthChecks` arrays afterwards. Operations with
 * no configured response return `{ error: "not_configured" }` so tests
 * must be explicit about the paths they exercise.
 */
export class GralkorInMemoryClient implements GralkorClient {
  private responses = new Map<Op, Result<unknown>>();

  readonly recalls: Array<[string, string, string]> = [];
  readonly captures: Array<[string, string, Message[]]> = [];
  readonly searches: Array<[string, string, string]> = [];
  readonly adds: Array<[string, string, string | null]> = [];
  readonly endSessions: Array<[string]> = [];
  readonly healthChecks: Array<[]> = [];
  readonly indicesBuilds: Array<[]> = [];
  readonly communitiesBuilds: Array<[string]> = [];

  setResponse(op: Op, response: Result<unknown>): void {
    this.responses.set(op, response);
  }

  reset(): void {
    this.responses.clear();
    this.recalls.length = 0;
    this.captures.length = 0;
    this.searches.length = 0;
    this.adds.length = 0;
    this.endSessions.length = 0;
    this.healthChecks.length = 0;
    this.indicesBuilds.length = 0;
    this.communitiesBuilds.length = 0;
  }

  private respond<T>(op: Op): Promise<Result<T>> {
    const r = this.responses.get(op) ?? { error: "not_configured" };
    return Promise.resolve(r as Result<T>);
  }

  async recall(groupId: string, sessionId: string, query: string): Promise<Result<string | null>> {
    this.recalls.push([groupId, sessionId, query]);
    return this.respond("recall");
  }

  async capture(sessionId: string, groupId: string, messages: Message[]): Promise<Result<true>> {
    this.captures.push([sessionId, groupId, messages]);
    return this.respond("capture");
  }

  async endSession(sessionId: string): Promise<Result<true>> {
    this.endSessions.push([sessionId]);
    return this.respond("endSession");
  }

  async memorySearch(groupId: string, sessionId: string, query: string): Promise<Result<string>> {
    this.searches.push([groupId, sessionId, query]);
    return this.respond("memorySearch");
  }

  async memoryAdd(
    groupId: string,
    content: string,
    sourceDescription: string | null,
  ): Promise<Result<true>> {
    this.adds.push([groupId, content, sourceDescription]);
    return this.respond("memoryAdd");
  }

  async healthCheck(): Promise<Result<true>> {
    this.healthChecks.push([]);
    return this.respond("healthCheck");
  }

  async buildIndices(): Promise<Result<{ status: string }>> {
    this.indicesBuilds.push([]);
    return this.respond("buildIndices");
  }

  async buildCommunities(
    groupId: string,
  ): Promise<Result<{ communities: number; edges: number }>> {
    this.communitiesBuilds.push([groupId]);
    return this.respond("buildCommunities");
  }
}
